### ISSUE TITLE
Limit certificate type

### ISSUE THREAD

**Issue by sbernard31** [eclipse/californium#752]
Currently when you set a X509 credentials, scandium uses `*_certificate_type` extension to express it support `X.509 (0x00)` or `Raw Public Key  (0x02`.

You can eventually change the order by using `preferRawPublicKeys` in `DtlsConnectorConfig.Builder.setIdentity()`.

I think all of this makes sense.

But still for LWM2M, I would like to limit my client to X.509 and so only send the X.509 `client_certificate_type`.

I currently use a workaround by using a `rpktruststore` which trusts nobody.

Maybe a solution could be to replace `preferRawPublicKeys` by an enum.
(`preferX509`, `preferRPK`, `onlyX509`, `onlyRPK`) ?

**Comment by boaks** [comment]
My only concern is, that since weeks now, I would like to "reformat" the logging in scandium, but I was help up by my own changes :-).  So, I need to wait a little more.

**Comment by boaks** [comment]
Should the (preferX509, preferRPK, onlyX509, onlyRPK) then only affect the client's list or
the server's list as well?
The "onlyRPK" seems to be obsolete, for that case `setIdentity(PrivateKey, PublicKey)` without the `setIdentity(PrivateKey, Certificate[], boolean or enum)`is used.

I also stumbled over:

According https://tools.ietf.org/pdf/rfc7250.pdf#7, page 6, Figure 4, the client's certificate type is a list, the server's a single entry.

Later on page 8,

>    The client_certificate_type and server_certificate_type extensions
>    sent in the client hello each carry a list of supported certificate
>    types, sorted by client preference.  When the client supports only
>    one certificate type, it is a list containing a single element.
> 

Scandiums implementation uses a list of also for the server's certificate type.

I didn't check for erratas nor did I dive deep into it.

**Comment by sbernard31** [comment]
> Should the (preferX509, preferRPK, onlyX509, onlyRPK) then only affect the client's list or
the server's list as well?

I was thinking both client and server.

> The "onlyRPK" seems to be obsolete, for that case setIdentity(PrivateKey, PublicKey) without the setIdentity(PrivateKey, Certificate[], boolean or enum)is used.

`onlyRPK` was for people who provide their pubKey using X509 certificate format but I agree it is maybe a bit weird.

**Comment by boaks** [comment]
The current API and implementation seems to "support always RPK". 
If it should be changed to a selective usage, I would prefer to go for:

```
setRawPublicKeyIdentity(PrivateKey privateKey, PublicKey publicKey)
setRawPublicKeyIdentity(PrivateKey privateKey, Certificate... certificate)
setX509Identity(PrivateKey privateKey, Certificate[] certificateChain)
```

If both, RPK and X509 are used, then RawPublicKey is preferred by default and could be adopted with
`setPreferSendRawPublicKey(boolean)`. 

```
setRawPublicKeyTrustStore(TrustedRpkStore store)
setX509TrustStore(Certificate[] trustedCerts)
setX509CertificateVerifier(CertificateVerifier verifier)
``` 

The default value is then changed to "don't trust X509" or "don't trust RPK".
If bot are used, then RawPublicKey is preferred by default and could be adopted with
`setPreferAcceptRawPublicKey(boolean)`. 

That is an API break, but for me it seems to be more "straight forward".

**Comment by sbernard31** [comment]
I'm not against breaking the API, but I feel strange that I need to do something like that

setx509Identity(myprivatekey, mycertificate)
setRawPublicKey(myprivatekey, mycertificate.getPublicKey)

If I want to use the same private and public key for both RPK and X509.

**Comment by boaks** [comment]
`setRawPublicKey(myprivatekey, mycertificate.getPublicKey)`

I used `Certificate...` so it could be called with

`setRawPublicKey(myprivatekey, mycertificate)`
 
But I guess, you fell strange about the two calls and not the parameter.

For me, this API is direct, if I want RPK, setup the rpk identity, if I want to trust RPK, setup the trust. If x509 is wanted, setup the x509 identity, if x509 is trusted setup the x509 trust.
For me, that's is really simple.

**Comment by sbernard31** [comment]
> But I guess, you fell strange about the two calls and not the parameter.

You guessed right.

>  if I want RPK, setup the rpk identity, if I want to trust RPK, setup the trust. If x509 is wanted, setup the x509 identity, if x509 is trusted setup the x509 trust.

I rather agree with that. The only strange case is where you set both. I don't know if this is a common use case. I could imagine a use case where you have a server, you define an X509 certificate/private key without client authentication. A client connect to it. It don't care about checking server certificate and only support RPK. In that case this make sense for the server to automatically accept both RPK and X509.

But this looks like more Web/Https use cases, so maybe not adapted to CoAP.

**Comment by boaks** [comment]
So, what do you prefer? 
Should I modify the boolean of setIdentity to a enum?
Or adapt the config according my proposal?

**Comment by sbernard31** [comment]
I have no strong opinion, I just share my concern with you, trying to help to take a good decision.

Maybe another solution could be.

```java
setRPKIdentity(PrivateKey, PublicKey)  // RPK ONLY => "straight forward"
setX509Identity(PrivateKey, Certificate ...) // X509 ONLY => "straight forward"
// for less straight forward use case :
setIdentity(Mode, PrivateKey, Certificate) // Mode (preferX509, preferRPK, onlyX509, onlyRPK)
```
We can also rename as you proposed :
```java
setRawPublicKeyTrustStore(TrustedRpkStore store)
setX509TrustStore(Certificate[] trustedCerts)
setX509CertificateVerifier(CertificateVerifier verifier)
```

**Comment by boaks** [comment]
After rethinking a lot:

```
setIdentity(PrivateKey privateKey, PublicKey publicKey)
setIdentity(PrivateKey privateKey, Certificate[] certificateChain, CertificateType... certificateTypes)
```

I use a `CertificateType... certificateTypes` to be more flexible for new/other certificate types.

I added a `setRpkTrustAll()`, but it's not longer implicitly used for RPK, it must be set.

I found, that the usage of the CertificateType list gets more static with that configuration.
Therefore I changed from List to array. And I cleaned up the CertificateTypeExtensions processing, e.g. removed a X_509 only list from the CLIENT_HELLO.

So, a lot of depended small changes.
